std/experimental/safeint.d

```diff
@@ -0,0 +1,887 @@
+/** This $(D module) holds functions to safely compare integers of different
+sizes and signeness without any unwanted implicit casts.
+
+Additionally, this $(D module) implements a $(D SafeInt!T) type.
+This type has an explicit $(D nan) value, checks binary operation for over-
+and underflows, checks division by zero, and checks if values to assign can be
+represented by $(D T) where $(D T) is any integer type.
+*/
+module std.experimental.safeint;
+
+import std.traits : isFloatingPoint, isIntegral, isUnsigned, isNumeric,
+    isSigned, Unqual;
+import std.typetuple : TypeTuple;
+
+@safe pure:
+
+/** This function compares two integer of arbitrary type for equality.
+
+This function makes sure no implicit value propagation falsifies the result of
+the comparison.
+
+Params:
+    t = an integer value
+    s = an integer value
+
+Returns: $(D true) if the value of $(D t) is equal to the value of $(D s),
+    false otherwise.
+*/
+bool equal(T, S)(in T t, in S s) @nogc nothrow if (isIntegral!T && isIntegral!S)
```

Comment:
Even so, if people see `in` used in Phobos they might think it's a good idea to use it everywhere.